internal/release/promote/perform.go

```diff
@@ -1,98 +1,183 @@
+	targetEnv := opts.list.Environments[1]
 
-	sourceEnv := params.list.Environments[0]
-	targetEnv := params.list.Environments[1]
+	info := &PromotionInfo{
+		SourceEnvironment: sourceEnv,
+		TargetEnvironment: targetEnv,
+	}
 
-	for _, crossRelease := range params.list.SortedCrossReleases() {
-		// Skip releases already in sync
+	var promotedFiles []string
+	for _, crossRelease := range opts.list.SortedCrossReleases() {
 		promotedFile := crossRelease.PromotedFile
 		if promotedFile == nil {
 			continue
 		}
-		promotedReleaseNames = append(promotedReleaseNames, crossRelease.Name)
+		promotedFiles = append(promotedFiles, promotedFile.Path)
 
-		// Update target release file
 		sourceRelease := crossRelease.Releases[0]
 		targetRelease := crossRelease.Releases[1]
 		isCreatingTargetRelease := targetRelease == nil
 
 		p.promptProvider.PrintUpdatingTargetRelease(targetEnv.Name, crossRelease.Name, promotedFile.Path, isCreatingTargetRelease)
 
-		if err := p.yamlWriter.Write(promotedFile); err != nil {
-			return "", fmt.Errorf("writing release %q promoted target yaml to file %q: %w", crossRelease.Name, promotedFile.Path, err)
+		if opts.dryRun {
+			fmt.Printf("ℹ️ Dry-run: skipping writing promoted release %s to: %s\n", style.Resource(crossRelease.Name), style.SecondaryInfo(promotedFile.Path))
+		} else {
+			if err := p.yamlWriter.Write(promotedFile); err != nil {
+				return "", fmt.Errorf("writing release %q promoted target yaml to file %q: %w", crossRelease.Name, promotedFile.Path, err)
+			}
 		}
 
-		promotedFiles = append(promotedFiles, promotedFile.Path)
-
-		// Determine release-specific message
-		message := getPromotionMessage(crossRelease.Name, sourceRelease, targetRelease)
-		messages = append(messages, message)
+		releaseInfo := getReleaseInfo(sourceRelease, targetRelease, opts)
+		info.Releases = append(info.Releases, releaseInfo)
+		info.Error = errors.Join(info.Error, releaseInfo.Error)
 	}
 
-	// Nothing promoted?
 	if len(promotedFiles) == 0 {
 		return "", fmt.Errorf("no releases promoted, should not reach this point")
 	}
 
-	// Create new branch and commit and push changes
-	branchName := getBranchName(sourceEnv.Name, targetEnv.Name, promotedReleaseNames)
-	message := getCommitMessage(sourceEnv.Name, targetEnv.Name, promotedReleaseNames, messages)
-
-	err := p.gitProvider.CreateAndPushBranchWithFiles(branchName, promotedFiles, message)
+	commitTemplate := cmp.Or(opts.commitTemplate, defaultCommitAndPRTemplate)
+	commitMessage, err := renderMessage(commitTemplate, info)
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("rendering commit message: %w", err)
 	}
 
-	p.promptProvider.PrintBranchCreated(branchName, message)
+	branchName := getBranchName(info)
+	if opts.dryRun {
+		fmt.Printf("ℹ️ Dry-run: skipping creation of branch %s\nFiles:\n%s\nCommit message:\n%s\n",
+			style.Resource(branchName), style.SecondaryInfo("- "+strings.Join(promotedFiles, "\n- ")),
+			style.SecondaryInfo(commitMessage))
```

Comment:
Perfect. Let's add this in a comment in the `joy.yaml` in https://github.com/nestoca/catalog/pull/980 on how to preview the template.

Reply:
Good point, I just added that comment:
```
#
# Templates for `joy release promote` Pull Requests and Commits
# 
# When customizing those templates, you can use a command similar to the following to quickly iterate and test rendering locally:
# $ joy release promote --dry-run --skip-catalog-update --draft --no-prompt --source qa --target demo release1
#
```